### DISCUSSION TITLE
Lossless encoding on multi layer AVIF images

### DISCUSSION THREAD

**Discussion by Foxify52** [bigcat88/pillow_heif#307] [General]
The following is an encoder i wrote to take advantage of the multi layer support avif has. Normally when using this code to encode data by itself in the first layer, it can encode and decode perfectly without any issue. as soon as you try and encode data and place it in a layer, it seems to ignore the lossless encoding i specified for those layers. I dont think its an issue with the package as much as a user error but it is one of those edge cases that might now have been accounted for. Any help would be appreciated.

Encoder:
```import numpy as np
from PIL import Image
import pillow_heif
import zlib

pillow_heif.register_avif_opener()


def encode_to_image(data):
    binary_data = zlib.compress(data)
    side_length = int(np.ceil(np.sqrt((len(binary_data) + 3) // 4)))
    binary_array = np.pad(
        np.frombuffer(binary_data, dtype=np.uint8),
        (0, side_length**2 * 4 - len(binary_data)),
        mode="constant",
    ).reshape((side_length, side_length, 4))
    return Image.fromarray(binary_array, "RGBA")


base = Image.open("base.jpg")
encoded_images = [encode_to_image(b"Some binary data to encode")]
# Save the image
base.save(
    "output.avif",
    format="AVIF",
    save_all=True,
    primary_index=0,
    append_images=encoded_images,
    quality=-1,  # Required for lossless compression
    chroma=444,  # Required for correct color subsampling
    matrix_coefficients=0,  # Required to convert to RGB color space
)
```
Decoder:
```import zlib

import numpy as np
import pillow_heif
from PIL import Image, ImageSequence

pillow_heif.register_avif_opener()

img = Image.open("output.avif")
bin_data = ImageSequence.Iterator(img)[1].copy()
reshaped_binary_array = np.array(bin_data)
flattened_binary_array = reshaped_binary_array.flatten()
binary_data_with_padding = flattened_binary_array.tobytes()
binary_data_without_padding = zlib.decompress(binary_data_with_padding.rstrip(b"\x00"))
original_binary_data = b"Some binary data to encode"
print(f"Original binary data: {original_binary_data}")
print(f"Decoded binary data: {binary_data_without_padding}")
print(
    f"Does the decoded data match the original data? {binary_data_without_padding == original_binary_data}"
)
```

**Comment by bigcat88** [comment]
does this code works with any other format that Pillow natively supports?

**Reply by bigcat88** [reply]
Nevermind, looks like `matrix_coefficients` does not applied to the second image

**Comment by Foxify52** [comment]
Should i open an issue? like i said, definitely seems like one of those edge cases that likely wasn't in the original scope of the package.

**Comment by bigcat88** [comment]
As `matrix_coefficients` can be specified/applied on the image level and not on the `encoder` level for the whole file, it is a bug or a missing feature of the `pillow-heif` package.

I would prefer to implement it as an option as suggested here with the help of `second_im.encoderinfo`:

https://github.com/python-pillow/Pillow/pull/8483


But in your example you have bytes as input and not an PIL image...

**Reply by bigcat88** [reply]
I created an issue from this topic(https://github.com/bigcat88/pillow_heif/issues/308), this feature is not difficult to add, we just need to decide how to specify this parameter.
